crates/client/sync/src/commitments/mod.rs

```diff
@@ -144,33 +53,67 @@ pub fn calculate_state_root(contracts_trie_root: Felt, classes_trie_root: Felt)
 ///
 ///
 /// The updated state root as a `Felt`.
-pub fn csd_calculate_state_root(backend: &DeoxysBackend, csd: CommitmentStateDiff, block_number: u64) -> Felt {
-    let CommitmentStateDiff {
-        address_to_class_hash,
-        address_to_nonce,
-        storage_updates,
-        class_hash_to_compiled_class_hash,
-    } = csd;
+pub fn compute_state_root(backend: &DeoxysBackend, state_diff: &StateDiff, block_number: u64) -> Felt {
+    let StateDiff {
+        storage_diffs,
+        deprecated_declared_classes: _,
+        declared_classes,
+        deployed_contracts,
+        replaced_classes,
+        nonces,
+    } = state_diff;
+
     // Update contract and its storage tries
     let (contract_trie_root, class_trie_root) = rayon::join(
         || {
-            contract_trie_root(backend, address_to_class_hash, address_to_nonce, storage_updates, block_number)
+            contract_trie_root(backend, deployed_contracts, replaced_classes, nonces, storage_diffs, block_number)
                 .expect("Failed to compute contract root")
         },
-        || {
-            class_trie_root(backend, class_hash_to_compiled_class_hash, block_number)
-                .expect("Failed to compute class root")
-        },
+        || class_trie_root(backend, declared_classes, block_number).expect("Failed to compute class root"),
     );
+
     calculate_state_root(contract_trie_root, class_trie_root)
 }
 
+/// Compute the root hash of a list of values.
+// The `HashMapDb` can't fail, so we can safely unwrap the results.
+pub fn compute_root<H>(values: &[Felt]) -> Felt
```

Comment:
`values` could be an iterator.
Would avoid needless collect from the callers

Reply:
You are right that using an iterator could avoid needless collections.
In most callers, we use `par_iter` from rayon to benefit from parallel processing.
However, rayon produces parallel iterators (ParallelIterator), which are not directly compatible with standard iterators (Iterator).
Additionally, methods like enumerate are not directly available on ParallelIterator, further complicating the direct use of iterators in this context.